### ISSUE TITLE
Problem: Clearing env vars breaks workflows

### ISSUE THREAD

**Issue by umglurf** [ansible-semaphore/semaphore#2464]
### Issue

[This security fix](https://github.com/semaphoreui/semaphore/commit/f33944e0429f711b24e03600a0c1dab7460b3a2d) breaks many legitimate workflows by removing every single environment variable before running a pipeline.
I have several environment variables that I have set that I need to pass on the ansible run. On of them is for instance `HTTPS_PROXY`, now the pipelines are unable to reach out because this is no longer passed on.
Instead of removing every environment variable, would it be possible to use a user configurarable whitelist?

### Impact

Ansible (task execution)

### Installation method

Docker

### Database

Postgres

### Semaphore Version

2.10.32-f33944e-1729509092

**Comment by Omicron7** [comment]
Just coming on to report this as well. We use quite a few ENV vars set in the the docker container running Semaphore, and inventory, and playbooks are no longer working as they require the ENV vars for authentication and configuration.

**Comment by vdudejon** [comment]
Reporting as well, closed my own issue #2467 as a duplicate.  Reverting to v2.10.31 resolved the problem.

**Comment by pascaliske** [comment]
This breaks for me as well – I have a few environment variables for proxy settings and installing private galaxy roles configured...

**Comment by fiftin** [comment]
Added env vars to config file. PR: https://github.com/semaphoreui/semaphore/pull/2470

Is is solution the issue?

**Comment by umglurf** [comment]
> Added env vars to config file. PR: #2470
> 
> Is is solution the issue?

It would solve the issue, but as commented in the PR by @Omicron7, having a whitelist of environment variables would be preferable.

**Comment by macau23** [comment]
Could this breaking change be made in a major version?

Also could the path include /bin? (See https://fedoraproject.org/wiki/Changes/Unify_bin_and_sbin)

**Comment by mlanner** [comment]
I've had some similar issues as described here, although I'm not convinced they are technically caused by this change, but they are nonetheless undesirable. Like @macau23 mentions, the paths included by Semaphore seems to be to look for executables in `:/sbin:/usr/sbin:/usr/local/sbin`, which is not typically where they tend to be installed. For example, in the error message below:

```
fatal: [server01]: FAILED! => {"changed": false, "msg": "Failed to find required executable \"rsync\" in paths: :/sbin:/usr/sbin:/usr/local/sbin"}
```

... Semaphore is looking in the "wrong" place for rsync. If I run the same playbook from the command line, Ansible properly finds rsync in `/usr/bin/rsync`, where it indeed is located. So, my conclusion is that this is a Semaphore issue.

By simply putting a symlink in `/sbin` to `/usr/bin/rsync` I was able to work around this problem, but I consider that more of a hack/workaround than a real solution. 

Again, this might not actually be directly related to this issue, and if so, please let me know. Maybe there's a way to tell Semaphore what the path should be and I just missed it? But, in any case, I think `/sbin` might not be the best default value. I can also create a new issue to track the behavior I outlined above.

**Comment by steadfasterX** [comment]
yes that got me as well today.. @mlanner your workaround is not a solution ofc and I did that as well at first (for the git binay) but the problem lays deeper and is discussed in PR: [#2470](https://github.com/semaphoreui/semaphore/pull/2470) . it also breaks completely if you use e.g. AWS modules requiring to set a specific PYTHONPATH etc.

so what I used before 2.10.32 was:

```
[Unit]
Description=Ansible Semaphore
Documentation=https://docs.ansible-semaphore.com/
Wants=network-online.target
After=network-online.target
ConditionPathExists=/usr/bin/semaphore
ConditionPathExists=/etc/semaphore/config.json

[Service]
User=semaphore
Group=semaphore
ExecStart=/usr/bin/semaphore service --config /etc/semaphore/config.json
ExecReload=/bin/kill -HUP $MAINPID
Restart=always
RestartSec=20s
# add path so the executables are found
Environment=PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:~/.local/bin"
# to fix "python module not found" issues, the following WorkingDirectory must point to the same
# directory as ` tmp_path`  variable in your semaphore config.json
# if using non-default ansible.cfg params, it also should contain your `.ansible.cfg` (yes, `dot`ansible.cfg)
WorkingDirectory=/opt/data/semaphore/plays
# besides the WorkingDirectory we need to tell systemd where to find your pip installed libs:
# to find out the user's as $User exec: python -c "import site; print(site.USER_SITE)"
Environment="PYTHONPATH=/home/semaphore/.local/lib/python3.11/site-packages"
# pass environment vars to child processes:
PassEnvironment=PYTHONPATH PATH

[Install]
WantedBy=multi-user.target
```

downgrading to 2.10.31 "solved" that issue until we found a proper solution. 

also (as suggested by @macau23 already) maybe such breaking changes should better happen in a major release and more clear described what might break? I had read "Clear run app's environment" and wasn't aware about the big impact it would have tbh

**Comment by fiftin** [comment]
Please try v2.10.33.

Fixed issue with `PATH` environment variable. 

Added options `SEMAPHORE_ENV_VARS` and `SEMAPHORE_FORWARDED_ENV_VARS`.

**Comment by umglurf** [comment]
Hi @fiftin 

I can verify that 2.10.33 using `SEMAPHORE_FORWARDED_ENV_VARS` works for me, thank you

**Comment by pascaliske** [comment]
@fiftin Thank you - `SEMAPHORE_FORWARDED_ENV_VARS` works perfectly fine for my use case!

**Comment by mlanner** [comment]
Sorry if I'm being dense here, but where should I add the two options `SEMAPHORE_ENV_VARS` and `SEMAPHORE_FORWARDED_ENV_VARS`? In a non-Docker environment, I'm assuming into the `config.json`, like so:

```
{
   "env_vars": {"SOME_VAR": "test"},
   "forwarded_env_vars": ["VAR1", "VAR2"],
}
```

And, as an example, `VAR1` could be `PATH`, as discussed here https://github.com/semaphoreui/semaphore/pull/2470#issuecomment-2429482407 ?

Thanks in advance.

**Comment by umglurf** [comment]
> Sorry if I'm being dense here, but where should I add the two options `SEMAPHORE_ENV_VARS` and `SEMAPHORE_FORWARDED_ENV_VARS`? In a non-Docker environment, I'm assuming into the `config.json`, like so:
> 
> ```
> {
>    "env_vars": {"SOME_VAR": "test"},
>    "forwarded_env_vars": ["VAR1", "VAR2"],
> }
> ```

Yes, you can add it to config.json as you show. If you run it in a systemd service, you can also set the environment variables in the service if that works better
```
[Service]
Environment=SEMAPHORE_FORWARDED_ENV_VARS=[\"VAR1\", \"VAR2\"]
```

> 
> And, as an example, `VAR1` could be `PATH`, as discussed here [#2470 (comment)](https://github.com/semaphoreui/semaphore/pull/2470#issuecomment-2429482407) ?
> 

https://github.com/semaphoreui/semaphore/commit/77358e3166cfb19d81d6ba2ad9ba9c03377a39b9 which is included in 2.10.34 passes on PATH as is, so with that there is no need to add PATH to forwarded env vars.
